File: client-java/src/main/java/io/maxi/examples/client/Player.java
package io.maxi.examples.client;

import io.maxi.annotation.MaxiField;
import io.maxi.annotation.MaxiType;

/**
 * Player DTO — mirrors the P:Player type in shared/sports.mxs.
 * Annotated with {@link MaxiType}/{@link MaxiField} so that
 * {@link io.maxi.api.SchemaRegistry#buildTypeDef} can derive the MAXI
 * schema from the class instead of the hand-written builder method.
 */
@MaxiType(alias = "P", name = "Player")
public class Player {

    @MaxiField(typeExpr = "int")
    public int id;

    @MaxiField(required = true)
    public String name;

    @MaxiField(typeExpr = "enum[forward,midfielder,defender,goalkeeper]")
    public String position;

    @MaxiField(typeExpr = "int")
    public int birthYear;

    @MaxiField(typeExpr = "T")
    public Team team;

    public Player() {}
}

File: server-php/src/Model/Player.php
<?php

declare(strict_types=1);

namespace App\Model;

use Maxi\Attribute\MaxiField;
use Maxi\Attribute\MaxiType;

/**
 * Player DTO — mirrors the P:Player type in shared/sports.mxs.
 *
 * Annotated with #[MaxiType] / #[MaxiField] so that MaxiSchemaRegistry
 * can derive the MAXI schema from the class rather than from hand-written
 * array descriptors.
 */
#[MaxiType(alias: 'P', name: 'Player')]
class Player
{
    public function __construct(
        #[MaxiField(typeExpr: 'int')]
        public int $id = 0,

        #[MaxiField(required: true)]
        public string $name = '',

        #[MaxiField(typeExpr: 'enum[forward,midfielder,defender,goalkeeper]')]
        public string $position = '',

        #[MaxiField(typeExpr: 'int')]
        public int $birthYear = 0,

        #[MaxiField(typeExpr: 'T')]
        public ?Team $team = null,
    ) {}
}
